src/sentry/digests/backends/redis.py

```diff
@@ -0,0 +1,442 @@
+from redis.client import Script
+from redis.exceptions import (
+    ResponseError,
+    WatchError,
+)
+
+from sentry.digests import (
+    Record,
+    ScheduleEntry,
+)
+from sentry.digests.backends.base import Backend
+from sentry.utils.cache import Lock
+
+
+logger = logging.getLogger('sentry.digests')
+
+SCHEDULE_PATH_COMPONENT = 's'
+SCHEDULE_STATE_WAITING = 'w'
+SCHEDULE_STATE_READY = 'r'
+
+TIMELINE_PATH_COMPONENT = 't'
+TIMELINE_ITERATION_PATH_COMPONENT = 'i'
+TIMELINE_DIGEST_PATH_COMPONENT = 'd'
+TIMELINE_RECORD_PATH_COMPONENT = 'r'
+
+
+def ilen(iterator):
+    i = 0
+    for i, _ in enumerate(iterator):
+        pass
+    return i
+
+
+def make_schedule_key(namespace, state):
+    return '{0}:{1}:{2}'.format(namespace, SCHEDULE_PATH_COMPONENT, state)
+
+
+def make_timeline_key(namespace, key):
+    return '{0}:{1}:{2}'.format(namespace, TIMELINE_PATH_COMPONENT, key)
+
+
+def make_iteration_key(timeline_key):
+    return '{0}:{1}'.format(timeline_key, TIMELINE_ITERATION_PATH_COMPONENT)
+
+
+def make_digest_key(timeline_key):
+    return '{0}:{1}'.format(timeline_key, TIMELINE_DIGEST_PATH_COMPONENT)
+
+
+def make_record_key(timeline_key, record):
+    return '{0}:{1}:{2}'.format(timeline_key, TIMELINE_RECORD_PATH_COMPONENT, record)
+
+
+# Ensures an timeline is scheduled to be digested.
+# KEYS: {WATING, READY}
+# ARGV: {TIMELINE, TIMESTAMP}
+ENSURE_TIMELINE_SCHEDULED_SCRIPT = """\
+-- Check to see if the timeline exists in the "waiting" set (heuristics tell us
+-- that this should be more likely than it's presence in the "ready" set.)
+local waiting = redis.call('ZSCORE', KEYS[1], ARGV[1])
+
+if waiting ~= false then
+    -- If the item already exists, update the score if the provided timestamp
+    -- is less than the current score.
+    if tonumber(waiting) > tonumber(ARGV[2]) then
+        redis.call('ZADD', KEYS[1], ARGV[2], ARGV[1])
+    end
+    return
+end
+
+-- Otherwise, check to see if the timeline already exists in the "ready" set.
+-- If it doesn't, it needs to be added to the "waiting" set to be scheduled.
+if redis.call('ZSCORE', KEYS[2], ARGV[1]) == false then
+    redis.call('ZADD', KEYS[1], ARGV[2], ARGV[1])
+    return
+end
+"""
+
+
+# Trims a timeline to a maximum number of records.
+# Returns the number of keys that were deleted.
+# KEYS: {TIMELINE}
+# ARGV: {LIMIT}
+TRUNCATE_TIMELINE_SCRIPT = """\
+local keys = redis.call('ZREVRANGE', KEYS[1], ARGV[1], -1)
+for i, record in pairs(keys) do
+    redis.call('DEL', KEYS[1] .. ':{TIMELINE_RECORD_PATH_COMPONENT}:' .. record)
+    redis.call('ZREM', KEYS[1], record)
+end
+return table.getn(keys)
+""".format(TIMELINE_RECORD_PATH_COMPONENT=TIMELINE_RECORD_PATH_COMPONENT)
+
+
+# XXX: Passing `None` as the first argument is a dirty hack to allow us to use
+# this more easily with the cluster
+ensure_timeline_scheduled = Script(None, ENSURE_TIMELINE_SCHEDULED_SCRIPT)
+truncate_timeline = Script(None, TRUNCATE_TIMELINE_SCRIPT)
+
+
+class RedisBackend(Backend):
```

Comment:
it would be nice to describe the redis data layout here as part of the docstring